fix: text selection two handles directionality.

muhammadkamel opened the pull request:
Fixes #187799

This PR fixes an issue where text selection handles would render facing the wrong direction when the text directionality differed from the ambient directionality (e.g., selecting English (LTR) text within an application set to Arabic (RTL)), and also when a single selection spans both LTR and RTL runs.

Previously, `TextSelectionOverlay` determined the start and end handle types based primarily on the ambient `TextDirection` and the logical selection range. This caused handles to swap (pointing away from the selection) in mixed-directionality scenarios.

This change updates the logic in `_updateSelectionOverlay` to derive each handle's type from the **per-endpoint `TextDirection`** reported by `TextSelectionPoint.direction`, falling back to the render object's `textDirection` when an endpoint does not provide one:

* Start handle: `TextDirection.ltr` → `TextSelectionHandleType.left`, `TextDirection.rtl` → `TextSelectionHandleType.right`.
* End handle: `TextDirection.ltr` → `TextSelectionHandleType.right`, `TextDirection.rtl` → `TextSelectionHandleType.left`.

Because each endpoint is evaluated independently, selections that cross a directionality boundary (e.g. `"abc مرحبا"`) get handle types that match the local direction at each end, so both handles continue to "hug" the selection.

On iOS, UIKit conventionally keeps selection handles aligned with the field's own direction rather than the per-endpoint direction. To preserve platform behavior, when `defaultTargetPlatform == TargetPlatform.iOS` both handles use the render object's `textDirection` instead of the endpoint directions.

When a non-collapsed selection yields fewer than two endpoints (e.g. the selection is scrolled or clipped out of view), handle direction falls back to the field's `textDirection` instead of asserting.

Collapsed selections continue to use `TextSelectionHandleType.collapsed` for both handles and are unaffected by this change.

**Tests added**

Added new tests in `text_selection_test.dart`:

Android (`TargetPlatformVariant.only(TargetPlatform.android)`), driven by a table of `_DirectionalityTestCase`s that assert the endpoint directions reported by `RenderEditable` and the handle types received by a spy `TextSelectionControls`:

* ambient LTR, English text
* ambient RTL, English text
* ambient RTL, Arabic text
* ambient LTR, Arabic text
* ambient LTR, mixed `"abc مرحبا"` (English then Arabic)
* ambient RTL, mixed `"abc مرحبا"` (English then Arabic)

iOS (`TargetPlatformVariant.only(TargetPlatform.iOS)`):

* Verifies that for the mixed `"abc مرحبا"` selection under ambient LTR, both handles follow the field's `textDirection` (i.e. `left` / `right`) rather than the per-endpoint directions — exercising the UIKit-compatibility branch.

The new directionality tests set `selectAllOnFocus: false` explicitly (with an inline comment) because web CI defaults `EditableText.selectAllOnFocus` to `true`, which can select all text on focus and prevent handle rebuilds in these tests.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

![image_issue](https://github.com/user-attachments/assets/7de61b60-c1dd-44bb-a5c8-34ff5d451350)

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

muhammadkamel committed:
fix: text selection directionality.

muhammadkamel committed:
refactor: introduce `_DirectionalityTestCase` class for improved test case definition and use `math.min`/`math.max`

muhammadkamel commented:
Hi @dkwingsmt,
Could you please assign a reviewer for this PR?

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -540,22 +540,41 @@ class TextSelectionOverlay {
+
+    if (!_selection.isCollapsed) {
+      if (endpoints.length == 2) {
+        if (endpoints[0].point.dx <= endpoints[1].point.dx) {
```
I think we can avoid this additional level of nesting here if we make this case a boolean, and use a ternary operator in the handle assignments below..

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -1899,6 +1901,132 @@ void main() {
+        final focusNode = FocusNode();
+
+        await tester.pumpWidget(
+          MaterialApp(
```
I think we want to avoid using `Material` in the widgets library moving forward (including `TextField`). However, in this case the infrastructure (helper test widgets) are not yet in place. I am working on a test text field https://github.com/flutter/flutter/pull/180494/ but it do not have a solution for testing selection handles yet. A test `WidgetsApp` is also being worked on here https://github.com/flutter/flutter/pull/180456. fyi @loic-sharma @justinmc

Renzo-Olivares reviewed (commented):
Thank you for the contribution @muhammadkamel, this change definitely helps improve the text selection experience.

I verified that this is also the behavior on native mobile platforms.

#### Jetpack Compose (Android) with RTL set for application.

<img width="1080" height="2364" alt="55" src="https://github.com/user-attachments/assets/85ae68f6-0797-4b56-b904-18b0f096d564" />


#### SwiftUI (iOS) with RTL set at application level

<img width="320" height="571" alt="Screenshot 2026-01-06 at 5 32 45 PM" src="https://github.com/user-attachments/assets/56ae8d68-ba33-4f2d-96f3-53f554ee2be2" />

WRT to RTL text in general I think this also improves the experience slightly but text selection is still a bit buggy in RTL when using the text selection handles. The magnifier also seems to be broken when the application level text directionality is RTL and the textfield is LTR. That can be improved in a separate PR.

muhammadkamel committed:
fix: simplify text selection handle type determination

muhammadkamel committed:
fix: improve comments for clarity in text selection tests

muhammadkamel committed:
fix: add punctuation for clarity in text selection test comments

muhammadkamel committed:
fix: enhance text selection handle type determination.

Renzo-Olivares commented:
Hi @muhammadkamel, just wanted to follow up if there is any updates on this PR I should review? During my review last week I was still observing issues https://github.com/flutter/flutter/pull/179928#discussion_r2683636141. Let me know if I should take a look again.

muhammadkamel commented:
> Hi @muhammadkamel, just wanted to follow up if there is any updates on this PR I should review? During my review last week I was still observing issues [#179928 (comment)](https://github.com/flutter/flutter/pull/179928#discussion_r2683636141). Let me know if I should take a look again.

Hi @Renzo-Olivares 
I will work on fixing this behavior issue. But I have a question, should the handle (Left) or (Right) direction changed base on the text or on the app locale (Directionality)?

muhammadkamel committed:
Fix selection width style for Arabic text and improve handle direction propagation:

1. Fix single-word Arabic text selection on Android using BoxWidthStyle.max
   when all text is selected, matching native Android behavior.
2. Add textDirection property to SelectionOverlay and TextSelectionOverlay
   to properly propagate text direction to selection handles.
3. Extract _getEffectiveSelectionWidthStyle() method for cleaner code.
4. Extract _resolveTextDirection() helper to reduce duplication in
   _buildStartHandle and _buildEndHandle.
5. Improve handle direction comments to clarify position vs pointing direction.
Tests added:
- Selection width style tests for Android (select all, partial selection)
- Selection width style test for iOS default behavior
- Arabic and English partial selection tests
- textDirection propagation test for handles

Renzo-Olivares commented:
Hi @muhammadkamel, thank you for your patience on this one. After confirming some of the native behavior on Jetpack compose and Android View API's, I think we are able to closely match the native behavior with your previous change in https://github.com/flutter/flutter/pull/179928/changes/7ff99718624081dec65a6dc9937f52fad3e69a8e . Checking the `TextDirection` of the `TextSelectionPoint` end points should be enough as you did in that commit. 

I was mistaken about the multi-line behavior before in https://github.com/flutter/flutter/pull/179928#discussion_r2683636141. Though I still think there may be a bug when moving the selection across multiple lines in the scenario where the App's directionality is RTL but the text is LTR, in the linked image the selection originally started on the first line on the word "world" but moving that selection to the line below, adds the "?" on the first line to the selection. This can be solved in a different PR though as the issue existed before this change.

I think some of the recent additional changes like passing the directionality to the `TextSelectionOverlay`, and `_SelectionHandleOverlay` are not needed. Can you explain why the changes for `getEndPointsForSelection` are needed and `EditableText`?

muhammadkamel committed:
fix: use endpoint TextDirection to determine selection handle types

Matches native Android/Compose behavior for mixed-directionality text by checking the text direction at each selection endpoint rather than using ambient text direction.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -1125,6 +1155,21 @@ class SelectionOverlay {
+  /// This is used to determine the direction of the selection handles.
+  ///
+  /// Changing the value while the handles are visible causes them to rebuild.
+  TextDirection? get textDirection => _textDirection;
```
This member and its equivalent in `TextSelectionOverlay` doesn't seem to be used by anything, is this needed?

Renzo-Olivares reviewed (commented):
Hi @muhammadkamel, thank you for the quick changes, I left a few comments. I'm also still seeing the changes to `getEndPointsForSelection` and `EditableText`.

muhammadkamel committed:
Simplify _updateSelectionOverlay

muhammadkamel committed:
Restore outer TapRegion for context menu with SelectableRegion

muhammadkamel committed:
Refactor text selection handling for mixed-directionality support on Android and iOS

muhammadkamel committed:
Replace MaterialApp with TestWidgetsApp

muhammadkamel committed:
Enhance text selection handling for mixed-directionality support in editable text

Renzo-Olivares commented on packages/flutter/lib/src/widgets/editable_text.dart:
```diff
@@ -2095,6 +2096,9 @@ class EditableText extends StatefulWidget {
       }
       return ui.BoxWidthStyle.tight;
     }
+    if (defaultTargetPlatform == TargetPlatform.android) {
```
This was recently changed in https://github.com/flutter/flutter/pull/167762 after an audit of the selection style on native platforms. Is this based on the behavior you are observing on Android?

muhammadkamel replied:
@Renzo-Olivares 
I will fix this issue in another PR

Renzo-Olivares replied:
fyi, this is already being fixed in https://github.com/flutter/flutter/pull/185239.

justinmc commented:
@muhammadkamel FYI you have a branch conflict to resolve here.

muhammadkamel commented:
> @muhammadkamel FYI you have a branch conflict to resolve here.

Hi @justinmc 

Thanks!
I have resolved the conflict.

muhammadkamel committed:
fix: Update text selection handling for mixed-directionality text and update related tests

muhammadkamel committed:
fix: Improve assertion message and refactor text selection handle direction logic

muhammadkamel commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -1748,15 +1748,45 @@ class RenderEditable extends RenderBox
       final Offset start = Offset(0.0, preferredLineHeight) + caretOffset + paintOffset;
       return <TextSelectionPoint>[TextSelectionPoint(start, null)];
     } else {
+      final int startOffset = selection.start;
```
You're right. I reverted the RenderEditable/getEndpointsForSelection changes.

justinmc commented:
@muhammadkamel Looks like there is another conflict (sorry for all the maintenance work here).

muhammadkamel commented:
> @muhammadkamel Looks like there is another conflict (sorry for all the maintenance work here).

No problem! Done!

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -1917,6 +1917,173 @@ void main() {
+      expectedStartHandleType: TextSelectionHandleType.left,
+      expectedEndHandleType: TextSelectionHandleType.left,
+    ),
+    _DirectionalityTestCase(
```
I think this last case should be:

```dart
expectedStartEndpointDirection: TextDirection.rtl,
expectedEndEndpointDirection: TextDirection.ltr,
expectedStartHandleType: TextSelectionHandleType.right,
expectedEndHandleType: TextSelectionHandleType.right,
```

muhammadkamel replied:
Good catch!

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -1917,6 +1917,173 @@ void main() {
+        final List<TextSelectionPoint> endpoints = renderEditable.getEndpointsForSelection(
+          controller.selection,
+        );
+        expect(endpoints.length, 2);
```
nit: consider `expect(endpoints, hasLength(2));` similar to the ios test.

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -2146,6 +2313,77 @@ class FakeTextSelectionDelegate extends Fake implements TextSelectionDelegate {
+    builtHandleTypes.clear();
+  }
+
+  TextSelectionHandleType get lastStartHandleType => builtHandleTypes[builtHandleTypes.length - 2];
```
I would just check `builtHandleTypes` directly. That way you can also do `expect(builtHandleTypes, hasLength(2))` before querying `builtHandleTypes.first/last`.

Renzo-Olivares reviewed (commented):
Hi @muhammadkamel, thank you for your patience on this one! Left a few small comments. Could you also update the PR description to more accurately fit the current implementation. There are also test/analyzer failures that need to be addressed.

muhammadkamel committed:
Update text selection test to correctly handle mixed directionality

Renzo-Olivares reviewed (commented):
Hi @muhammadkamel, thank you for the changes. There is still a comment that needs to be addressed https://github.com/flutter/flutter/pull/179928#discussion_r3256454929.

muhammadkamel committed:
Refactor text selection test to use hasLength and first/last for handle types

muhammadkamel commented:
> Hi @muhammadkamel, thank you for the changes. There is still a comment that needs to be addressed [#179928 (comment)](https://github.com/flutter/flutter/pull/179928#discussion_r3256454929).

Check now please

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -2146,6 +2313,73 @@ class FakeTextSelectionDelegate extends Fake implements TextSelectionDelegate {
+    VoidCallback? onTap,
+  ]) {
+    builtHandleTypes.add(type);
+    return SizedBox.square(key: ValueKey<TextSelectionHandleType>(type), dimension: textLineHeight);
```
I don't think this key is being used. I think we can just remove it.

Renzo-Olivares replied:
I think this comment was not resolved yet, we should remove the unused key.

Renzo-Olivares reviewed (dismissed):
LGTM with one final nit. Could you also update the PR description to more accurately fit the current implementation. This will need a secondary reviewer and they may be confused by the PR description since it does not match the current implementation.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -548,22 +548,52 @@ class TextSelectionOverlay {
+    final List<TextSelectionPoint> endpoints = renderObject.getEndpointsForSelection(_selection);
+    assert(endpoints.isNotEmpty);
+
+    late final TextSelectionHandleType startHandleType;
```
here and below: late final -> final

LongCatIsLooong reviewed (dismissed):
LGTM with nits.

muhammadkamel committed:
Replace late final with final in _updateSelectionOverlay

muhammadkamel committed:
Remove unused ValueKey from DirectionalitySpyTextSelectionControls

muhammadkamel commented:
> LGTM with one final nit. Could you also update the PR description to more accurately fit the current implementation. This will need a secondary reviewer and they may be confused by the PR description since it does not match the current implementation.

I've updated the PR description to reflect the current implementation.

Renzo-Olivares reviewed (dismissed):
LGTM, thank you for your contribution @muhammadkamel!

Renzo-Olivares commented:
@muhammadkamel, heads up you will need to fix the analyzer issues here. as well as any failing tests before this PR lands.

muhammadkamel commented:
> @muhammadkamel, heads up you will need to fix the analyzer issues here. as well as any failing tests before this PR lands.

Could you check now please.

muhammadkamel committed:
fix: resolve Linux analyze failures for text selection PR

Remove redundant bool type annotation to satisfy omit_obvious_local_variable_types
and apply dart format to text_selection_test.dart so CI analyze and format checks pass.

Renzo-Olivares commented:
Hi @muhammadkamel, there are still some failing tests that need to be addressed before landing this PR. Thank you!

muhammadkamel committed:
fix: avoid crash when non-collapsed selection has fewer than two endpoints

A non-collapsed selection can yield fewer than two endpoints when an
endpoint is clipped out of view (e.g. maxLength fields with a selection at
the text edge). The previous assert(endpoints.length == 2) turned this into
a fatal exception during _updateSelectionOverlay. Fall back to the render
object's textDirection for handle direction when fewer than two endpoints
are available.

muhammadkamel commented:
> Hi @muhammadkamel, there are still some failing tests that need to be addressed before landing this PR. Thank you!

Hi @Renzo-Olivares 
You're welcome! :)
Please check once again.

muhammadkamel committed:
Fix web directionality selection handle tests by disabling selectAllOnFocus in TestTextField.

muhammadkamel commented:
Hi @Renzo-Olivares 

Is there anything else needed from me to land this PR?

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -548,22 +548,47 @@ class TextSelectionOverlay {
+          defaultTargetPlatform == TargetPlatform.iOS;
+      final TextDirection startHandleDirection;
+      final TextDirection endHandleDirection;
+      if (preferRenderObjectDirectionForSelectionHandles || endpoints.length < 2) {
```
I would add a comment here on why a non-collapsed selection may have less than two endpoints.

Renzo-Olivares commented on packages/flutter/test/widgets/editable_text_tester.dart:
```diff
@@ -39,6 +39,7 @@ class TestTextField extends StatefulWidget {
     this.style,
     this.controller,
     this.onSubmitted,
+    this.selectAllOnFocus,
```
Is this necessary? `selectAllOnFocus` is false by default on mobile platforms.

muhammadkamel replied:
You're right that it's false by default on Android/iOS. We still need it because these tests also run on web CI, where the default is true. I'll default TestTextField to false so tests don't have to pass it explicitly.

Renzo-Olivares replied:
Looking at the test i'm curious why it failed in the first place. In the test we request the focus, after we clear the built handles and set a new selection. Shouldn't our new selection win in this case, or what was there a different point of failure?

muhammadkamel replied:
Good question — the explicit selection does win, but only when it actually changes the range.

On web, `selectAllOnFocus` defaults to `true`, so focus already selects the full text. For the mixed-directionality tests (and the iOS test), our intended selection is also the full range, so assigning it again is a no-op and no handles are rebuilt after `clearBuiltHandleTypes()`.

The partial-selection Android cases pass because focus selects all, then the test sets a shorter range, which triggers a rebuild. Defaulting `TestTextField` to `selectAllOnFocus: false` avoids the focus-time select-all so the explicit assignment always changes the selection.

Renzo-Olivares replied:
Sounds good to me, can you add a comment in the test next to `selectAllOnFocus` describing why it needs to be disabled.

muhammadkamel replied:
Done — removed the default from TestTextField so it inherits EditableText behavior. The new directionality tests now set selectAllOnFocus: false explicitly with an inline comment explaining the web CI behavior.

Renzo-Olivares reviewed (commented):
Hi @muhammadkamel, thank you for your patience on this one. I had a few small questions but otherwise looks good.

muhammadkamel committed:
Add comment explaining endpoints.length < 2 fallback in text selection overlay.

muhammadkamel committed:
Default TestTextField selectAllOnFocus to false for consistent test behavior.
Web CI defaults EditableText.selectAllOnFocus to true even when simulating
mobile platforms. Defaulting to false in the test helper removes the need for
call sites to pass it explicitly.

muhammadkamel committed:
Add screenshot for flutter/flutter#187799.

Renzo-Olivares commented:
Thank you for the changes @muhammadkamel, had a small question about the failing web tests https://github.com/flutter/flutter/pull/179928#discussion_r3398751730.

Renzo-Olivares reviewed (dismissed):
LGTM modulo small comment https://github.com/flutter/flutter/pull/179928#discussion_r3438272193. This PR will need to be rebased before landing.

muhammadkamel committed:
Document selectAllOnFocus default in TestTextField

muhammadkamel committed:
Remove selection handle screenshot asset

muhammadkamel commented:
Hi @Renzo-Olivares,

Could you please take another look when you have a chance?

muhammadkamel committed:
fix: expect max selection width style on Android

Renzo-Olivares commented on packages/flutter/test/widgets/editable_text_tester.dart:
```diff
@@ -39,6 +39,9 @@ class TestTextField extends StatefulWidget {
     this.onSubmitted,
+    // Keep test behavior consistent across platforms, including web CI where
+    // EditableText otherwise defaults selectAllOnFocus to true.
+    this.selectAllOnFocus = false,
```
Apologies, what I meant before is that this should not have a default, i.e. should inherit `EditableText` behavior. And in the tests we added we should set it to `false` explicitly and add a comment next to it as to why.

muhammadkamel replied:
Done — removed the default from TestTextField so it inherits EditableText behavior. The new directionality tests now set selectAllOnFocus: false explicitly with an inline comment explaining the web CI behavior.

muhammadkamel committed:
fix: set selectAllOnFocus explicitly in directionality tests

Remove the TestTextField default so it inherits EditableText behavior.
Only the new directionality tests opt out of selectAllOnFocus for web CI.

muhammadkamel committed:
fix: skip Android selection width test on web

On web, defaultSelectionWidthStyle returns tight for non-iOS
browsers regardless of defaultTargetPlatform.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-mixed-directionality

Renzo-Olivares commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -18794,6 +18802,60 @@ void main() {
     skip: kIsWeb,
   );
+
+  testWidgets(
```
I don't think the tests changes in `editable_text_test.dart` are related to the changes in this PR. If so we should probably revert/remove them.

Renzo-Olivares commented on packages/flutter/test/widgets/editable_text_tester.dart:
```diff
@@ -60,6 +60,11 @@ class TestTextField extends StatefulWidget {
+  /// Controls whether all text is selected when the field receives focus.
+  ///
+  /// When null, [EditableText] platform defaults apply.
+  final bool? selectAllOnFocus;
```
This is never set in the constructor. I think we accidentally removed it, we should add it back.

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -2088,6 +2088,179 @@ void main() {
+                controller: controller,
+                focusNode: focusNode,
+                selectionControls: customControls,
+                // Web CI defaults EditableText.selectAllOnFocus to true, which
```
I suggest something along the lines of, in this test and the other test where we set `selectAllOnFocus`:

```
On the web selectAllOnFocus defaults to true, interfering with this tests programmatic select all.
```

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-mixed-directionality

muhammadkamel committed:
Revert unrelated editable_text_test changes, wire TestTextField.selectAllOnFocus
through to EditableText, restore directionality test helpers dropped in the
master merge, and clarify the web selectAllOnFocus comments.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-mixed-directionality

muhammadkamel commented:
Hi @Renzo-Olivares 

Addressed your latest comments — please take another look when you get a chance.

muhammadkamel committed:
Merge remote-tracking branch 'upstream/master' into fix/text-selection-mixed-directionality

Renzo-Olivares commented:
@muhammadkamel can you try to rebase this to the latest `master`. Tests seem to be stuck at the moment.

muhammadkamel committed:
Merge branch 'master' into fix/text-selection-mixed-directionality

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -587,7 +615,7 @@ class TextSelectionOverlay {
   ///
   ///   * [spellCheckToolbarIsVisible], which is only whether the spell check menu
-  ///     specifically is visible.
+  ///     is visible.
```
We should revert this unrelated change.

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -548,22 +548,50 @@ class TextSelectionOverlay {
+          defaultTargetPlatform == TargetPlatform.iOS;
+      final TextDirection startHandleDirection;
+      final TextDirection endHandleDirection;
+      // A non-collapsed selection may still yield a single endpoint when
```
I wasn't able to write a test that actually reproduced this with a scrolled or clipped out of view selection.

I think the comment below would be more accurate.

```
// A non-collapsed selection might return fewer than two endpoints if the 
// text layout lacks boxes for the selected range. This typically happens when:
//
//  * Render lag: The overlay updated with a new editing value before the 
//    render object laid out the new text (selection offsets are out of bounds).
//  * Split graphemes: A selection boundary falls inside a multi-code-unit 
//    cluster (like an emoji or combining character).
//  * Degenerate layout: The layout is temporarily squashed (e.g., 
//    preferredLineHeight is 0 during a fold transition).
//
// In these cases, we fall back to the field's textDirection.
```

muhammadkamel committed:
Address review: drop unused spy handle key, restore toolbarIsVisible doc, and clarify endpoint-fallback comment.

muhammadkamel commented:
Hi @Renzo-Olivares — addressed your latest comments:

Removed the unused ValueKey from DirectionalitySpyTextSelectionControls (the spy already asserts via builtHandleTypes).
Reverted the unrelated toolbarIsVisible doc change.
Updated the < 2 endpoints fallback comment to your suggested wording (render lag / split graphemes / degenerate layout).
Please take another look when you have a chance.

Renzo-Olivares commented on packages/flutter/test/widgets/text_selection_test.dart:
```diff
@@ -2370,3 +2543,70 @@ class _MockTextSelectionHandleControls extends TextSelectionControls
   }
 }
+
+class DirectionalitySpyTextSelectionControls extends TextSelectionControls {
```
nit: I'd make this class private for now until it needs to be exposed to other tests.

muhammadkamel replied:
Check now!

Renzo-Olivares reviewed (dismissed):
LGTM w/ one small nit. Thank you for your patience on this one!

muhammadkamel committed:
Make DirectionalitySpyTextSelectionControls private in text selection tests.

Renzo-Olivares reviewed (approved):
RE-LGTM, thank you for the contribution @muhammadkamel!

muhammadkamel commented:
> RE-LGTM, thank you for the contribution @muhammadkamel!

You're welcome! :)

muhammadkamel committed:
Merge remote-tracking branch 'upstream/master' into fix/text-selection-mixed-directionality
